gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+  public void close()
+      throws IOException {
+    log.debug("Close called");
+    this.closer.close();
+  }
+
+  @Override
+  public void commit()
+      throws IOException {
+    log.debug("Commit called, will wait for commitTimeout : " + commitTimeoutInNanos / MILLIS_TO_NANOS + "ms");
+    long commitStartTime = System.nanoTime();
+    asyncDataWriter.flush();
+    while (((System.nanoTime() - commitStartTime) < commitTimeoutInNanos)  &&
+        (recordsIn.getCount() != (recordsSuccess.getCount() + recordsFailed.getCount())))
+    {
+      log.debug("Commit waiting... records produced: " + recordsIn.getCount() + " written: "
+          + recordsSuccess.getCount() + " failed: " + recordsFailed.getCount());
+      try {
+        Thread.sleep(commitStepWaitTimeMillis);
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted while waiting for commit to complete", e);
+      }
+    }
+    log.debug("Commit done waiting");
+    long recordsProducedFinal = recordsIn.getCount();
+    long recordsWrittenFinal = recordsSuccess.getCount();
+    long recordsFailedFinal = recordsFailed.getCount();
+    long unacknowledgedWrites  = recordsProducedFinal - recordsWrittenFinal - recordsFailedFinal;
+    long totalFailures = unacknowledgedWrites + recordsFailedFinal;
+    if (unacknowledgedWrites > 0) // timeout
+    {
+      log.warn("Timeout waiting for all writes to be acknowledged. Missing " + unacknowledgedWrites
+          + " responses out of " + recordsProducedFinal);
+    }
+    if (totalFailures > 0 && recordsProducedFinal > 0)
+    {
+      String message = "Commit failed to write " + totalFailures
+          + " records (" + recordsFailedFinal + " failed, " + unacknowledgedWrites + " unacknowledged) out of "
+          + recordsProducedFinal + " produced.";
+      double failureRatio = (double)totalFailures / (double)recordsProducedFinal;
+      if (failureRatio > failureAllowance)
+      {
+        message += "\nAborting because this is greater than the failureAllowance percentage: " + failureAllowance*100.0;
+        log.error(message);
+        throw new IOException(message);
+      }
+      else
+      {
+        message += "\nCommitting because failureRatio percentage: " + (failureRatio * 100.0) +
+            " is less than the failureAllowance percentage: " + (failureAllowance * 100.0);
+        log.warn(message);
+      }
+    }
+    log.info("Successfully committed " + recordsWrittenFinal + " records.");
+  }
+
+
+  public static AsyncWriterManagerBuilder builder() {
+    return new AsyncWriterManagerBuilder();
+  }
+
+  public static class AsyncWriterManagerBuilder {
+    private Config config = ConfigFactory.empty();
+    private long commitTimeoutInNanos = COMMIT_TIMEOUT_IN_NANOS_DEFAULT;
+    private long commitStepWaitTimeMillis = COMMIT_STEP_WAITTIME_MILLIS_DEFAULT;
+    private double failureAllowance = FAILURE_ALLOWANCE_DEFAULT;
+    private boolean retriesEnabled = RETRIES_ENABLED_DEFAULT;
```

Comment:
I'd rather have one setting (numRetries); numRetries == 0, means that retries are disabled.